### ISSUE TITLE
Netty Websocket stability discussion

### ISSUE THREAD

**Issue by firegnome** [bitrich-info/xchange-stream#80]
A month ago I committed an automatic reconnect for the poloniex websocket https://github.com/bitrich-info/xchange-stream/pull/43. This seems to work and I'm glad to see it integrated in the Netty class https://github.com/bitrich-info/xchange-stream/pull/56. This reconnect solves a lot of problems and is in my opinion the correct behaviour for the subscription.

Now we're still facing some stability problems with the websocket that aren't solved by the fix above.
1. The exchange drops the subscription on server side but doesn't inform us. (websocket stays open but no trades are sent)
2. A connection interrupt doesn't close the Netty websocket and doesn't trigger the disconnected event. 

My first idea was to enable tcp keepalive on the netty socket to recognize a connection lost, but this doesn't seem to solve the problem as described here: https://stackoverflow.com/questions/23238319/websockets-ping-pong-why-not-tcp-keepalive.

What I've learned is that we should use a custom ping pong but since poloniex doesn't provide such a functionality we have to use their own implementation. They solved this problem with heartbeats. @chuggol has also recognized this problem on gdax https://github.com/bitrich-info/xchange-stream/issues/66. 

Poloniex API heartbeats:
> In some markets, if there is no update for more than 1 second, a heartbeat message consisting of an empty argument list and the latest sequence number will be sent. These will go out once per second, but if there is no update for more than 60 seconds, the heartbeat interval will be reduced to 8 seconds until the next update.

What I ended up doing is to start a watcher thread that verifies that in the last 10 seconds at least a heartbeat or a trade/orderbook/... update was sent to verify that the websocket is still open. https://github.com/bitrich-info/xchange-stream/compare/develop...firegnome:develop#diff-0a517b95d1307fce62ef049280dd36c2R166

When the thread recognizes that the connection is lost it does following:
1. call the **resubscribe** function -> that will result in a connection timeout (I set the Netty websocket timeout to 10 seconds: https://github.com/bitrich-info/xchange-stream/compare/develop...firegnome:develop#diff-24429d40711776b7bfbbf1d0a497fd6eR118)
2. With this ReadTimeout Netty will recognize the lost connection and trigger the disconnected event
3. Now the reconnect/resubscribe procedure that was implemented in https://github.com/bitrich-info/xchange-stream/pull/43 will reestablish the connection.

I tested and improved this solution with the watcher thread and it really seems to work quite well and seems to solve the second problem. (I run the websocket 24/7) But so far I didn't create a pull request because it is in my opinion a really ugly fix that can't be easily adapted to other exchanges because its very specific. 

I look forward to your opinions and advice on this problem!

**Comment by the-foreigner** [comment]
I noticed the same issue, the trades eventually stop being sent while the socket remains  open. There is no error reported by the server. I haven't investigated further. This is with Bitstamp.

**Comment by GlebYarnykh** [comment]
Today i've got 

02:30:11,669 DEBUG [pusher-java-client timers] WebSocketConnection - Sending ping
02:30:14,870 DEBUG [Thread-2] WebSocketConnection - State transition requested, current [CONNECTED], new [RECONNECTING]
02:30:14,871 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to RECONNECTING from CONNECTED
02:30:15,879 DEBUG [pusher-java-client timers] WebSocketConnection - State transition requested, current [RECONNECTING], new [CONNECTING]
02:30:15,879 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to CONNECTING from RECONNECTING
02:30:16,715 DEBUG [pusher-java-client eventQueue] WebSocketConnection - State transition requested, current [CONNECTING], new [CONNECTED]
02:30:16,715 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to CONNECTED from CONNECTING

But after that my observables stopped getting any quotes. Is there any way to fix it?

Thanks

**Comment by firegnome** [comment]
Hi @GlebYarnykh 

Yes i think i've seen the same problem. And i guess i've figured out the cause.
This code caused the problem:
`connect().blockingAwait()`
It waits until the reconnect is completed. If you have no internet during this time it will hang forever there. I've added a timeout to prevent this:
`connect().blockingAwait(10, TimeUnit.SECONDS)`

Check out my newest commit 5d09ad5 to get theses changes. But notice that it's only working for poloniex2 at the time! 

If you want to test if its working, start the socket and cut the internet connection for a minute. After that the websocket should work normally again.

**Comment by firegnome** [comment]
I did test implement this on poloniex. The websocket watcher is only running on poloniex2. But the normal reconnect, when the server closes the connection, is also fixed for other exchanges.

You don't have to dispose or resubscribe anything this does all work automatic.

Which exchange do you use? Sadly there is no generic solution for all exchanges, so we need to implement this for every exchange.
I plan to do following:

- On sunday I'll create a pull request for the watcher thread in the Netty class.
- After that each exchange has to override a method that say whether the socket is active or needs to be reconnected.

I will implement this for poloniex, but it will be very easy to adapt this for other exchanges.

**Comment by nodarret** [comment]
My fix #89 , like @firegnome wrote, helps only when operating system have information that connection is closed. There are some scenarios when we don't know if connection is still alive. Most common solution is Ping/Heartbeat mechanism. Some exchanges, like poloniex have that mechanism but it is exchange specific.

**Comment by GlebYarnykh** [comment]
@firegnome 
The only problem I see is that when you reconnect, orderBooks become stale i.e. need to be reconnected and submitted full cycle (GDAX full channel for example or binance). Thus upper levels like StreamingMarketData service would NOT know about reconnect, and thus not know whether they should resubscribe/clear order books. Maybe my case is specific cause I use my own orderBooks structure on top of streaming service, but this might be a trouble for xchange-streaming as well in future..

**Comment by GlebYarnykh** [comment]
@firegnome 
Regarding Internet connection you mentioned. I have tried cutting internet for 5 minutes after receiving initial updates, and reconnect was not required really. Just messages with correct sequence number went through in a big burst and after that normally.

What timeout websocket should recieve to close it out on server side?

**Comment by trojanflash** [comment]
Really very strange reconnection implementation. When channelInactive called it does not mean that websocket connection was closed.

**Comment by firegnome** [comment]
@trojanflash From my point of view this is exactly what this method is ment for. 
In the javadoc https://netty.io/4.1/api/io/netty/channel/ChannelInboundHandler.html it does say following:

> channelInactive(ChannelHandlerContext ctx)The Channel of the ChannelHandlerContext was registered is now inactive and reached its end of lifetime.

From what I understand is this channel the websocket channel and not the subscribed channel on poloniex.

**Comment by trojanflash** [comment]
@firegnome You can test it on bitfinex, comment out reconnection code in NettyWebSocketClientHandler and wait some. Some time later channelInactive will be called, but messaged still receiving.

**Comment by firegnome** [comment]
@trojanflash Okay this sound like a very strange behaviour... Can you confirm that you have only one instance of BitfinexStreamingService? I just want to make sure that the channelInactive handler is not called from another instance, that was closed in the meantime.

**Comment by trojanflash** [comment]
@firegnome yeap, it was because of second instance, sorry. 
offtop
If you don't mind i write here, please add in future listener in NettyStreamingService:
`handler.handshakeFuture().addListener(f -> {
			    if(f.isSuccess()) {
				completable.onComplete();
			    } else {
				completable.onError(f.cause());
			    }
			});`
to produce an error if handshake failures.
thanks.

**Comment by firegnome** [comment]
> The only problem I see is that when you reconnect, orderBooks become stale i.e. need to be reconnected and submitted full cycle (GDAX full channel for example or binance).
@hlbkin You are totally right we need to think about some notify mechanisms form these cases.
@trojanflash I believe what you are proposing was already fixed by @hlbkin in #89

**Comment by daniel-cohen** [comment]
> Today i've got
> 
> 02:30:11,669 DEBUG [pusher-java-client timers] WebSocketConnection - Sending ping
> 02:30:14,870 DEBUG [Thread-2] WebSocketConnection - State transition requested, current [CONNECTED], new [RECONNECTING]
> 02:30:14,871 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to RECONNECTING from CONNECTED
> 02:30:15,879 DEBUG [pusher-java-client timers] WebSocketConnection - State transition requested, current [RECONNECTING], new [CONNECTING]
> 02:30:15,879 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to CONNECTING from RECONNECTING
> 02:30:16,715 DEBUG [pusher-java-client eventQueue] WebSocketConnection - State transition requested, current [CONNECTING], new [CONNECTED]
> 02:30:16,715 INFO [pusher-java-client eventQueue] PusherStreamingService - State changed to CONNECTED from CONNECTING
> 
> But after that my observables stopped getting any quotes. Is there any way to fix it?
> 
> Thanks

@hlbkin have you ever been able to solve this ? I've got a similar  issue today where "pusher" was stuck on reconnecting and just seem to have died:

```
2018-03-22 01:05:27.923 [pusher-java-client eventQueue] INFO  i.b.x.s.p.PusherStreamingService - State changed to RECONNECTING from CONNECTED
2018-03-22 01:05:28.932 [pusher-java-client eventQueue] INFO  i.b.x.s.p.PusherStreamingService - State changed to CONNECTING from RECONNECTING
2018-03-22 01:05:36.316 [pusher-java-client eventQueue] INFO  i.b.x.s.p.PusherStreamingService - State changed to RECONNECTING from CONNECTING
2018-03-22 01:05:40.319 [pusher-java-client eventQueue] INFO  i.b.x.s.p.PusherStreamingService - State changed to CONNECTING from RECONNECTING
```

**Comment by firegnome** [comment]
@daniel-cohen Are you running the version from my branch or from the main and which exchange do you use? We are still waiting, that #94 will be merged to the main branch. My trade bot is now running for months with my version so it seems to be very stable.

**Comment by daniel-cohen** [comment]
@firegnome No, it's not your branch. It's basically the version I forked from the original.
And I am using Bitstamp which is using Pusher.
